File: example/ts/utils.test.ts
// Copyright (c) 2026 Gibran Rodriguez <brangi000@gmail.com>
// SPDX-License-Identifier: MIT

import { describe, it, expect } from "vitest";
import { calculateAverage, fizzBuzz, reverseWords } from "./utils";

describe("calculateAverage", () => {
  // Happy paths
  it("should calculate the average of positive numbers", () => {
    expect(calculateAverage([1, 2, 3])).toBe(2);
  });

  it("should calculate the average of a single number", () => {
    expect(calculateAverage([42])).toBe(42);
  });

  it("should calculate the average with decimal results", () => {
    expect(calculateAverage([1, 2, 3, 4, 5])).toBe(3);
  });

  it("should handle large numbers", () => {
    expect(calculateAverage([1000000, 2000000, 3000000])).toBe(2000000);
  });

  it("should handle negative numbers", () => {
    expect(calculateAverage([-5, -3, -1])).toBe(-3);
  });

  it("should handle mixed positive and negative numbers", () => {
    expect(calculateAverage([-2, 0, 2])).toBe(0);
  });

  it("should handle fractional numbers", () => {
    expect(calculateAverage([0.5, 1.5, 2.5])).toBe(1.5);
  });

  // Edge cases
  it("should handle zero values", () => {
    expect(calculateAverage([0, 0, 0])).toBe(0);
  });

  it("should calculate average with two numbers", () => {
    expect(calculateAverage([10, 20])).toBe(15);
  });

  it("should handle very small decimal numbers", () => {
    expect(calculateAverage([0.1, 0.2, 0.3])).toBeCloseTo(0.2, 5);
  });

  // Error handling
  it("should throw an error for empty array", () => {
    expect(() => calculateAverage([])).toThrow(
      "Input must be a non-empty array"
    );
  });

  it("should throw an Error instance", () => {
    expect(() => calculateAverage([])).toThrow(Error);
  });
});

describe("fizzBuzz", () => {
  // Happy paths - fizzbuzz
  it("should return 'fizzbuzz' for multiples of 15", () => {
    expect(fizzBuzz(15)).toBe("fizzbuzz");
    expect(fizzBuzz(30)).toBe("fizzbuzz");
    expect(fizzBuzz(45)).toBe("fizzbuzz");
  });

  // Happy paths - fizz
  it("should return 'fizz' for multiples of 3 only", () => {
    expect(fizzBuzz(3)).toBe("fizz");
    expect(fizzBuzz(6)).toBe("fizz");
    expect(fizzBuzz(9)).toBe("fizz");
  });

  it("should return 'fizz' for 3", () => {
    expect(fizzBuzz(3)).toBe("fizz");
  });

  // Happy paths - buzz
  it("should return 'buzz' for multiples of 5 only", () => {
    expect(fizzBuzz(5)).toBe("buzz");
    expect(fizzBuzz(10)).toBe("buzz");
    expect(fizzBuzz(20)).toBe("buzz");
  });

  it("should return 'buzz' for 5", () => {
    expect(fizzBuzz(5)).toBe("buzz");
  });

  // Happy paths - number as string
  it("should return the number as a string for other numbers", () => {
    expect(fizzBuzz(1)).toBe("1");
    expect(fizzBuzz(2)).toBe("2");
    expect(fizzBuzz(4)).toBe("4");
    expect(fizzBuzz(7)).toBe("7");
  });

  // Edge cases
  it("should handle zero", () => {
    expect(fizzBuzz(0)).toBe("fizzbuzz");
  });

  it("should handle negative multiples of 15", () => {
    expect(fizzBuzz(-15)).toBe("fizzbuzz");
    expect(fizzBuzz(-30)).toBe("fizzbuzz");
  });

  it("should handle negative multiples of 3", () => {
    expect(fizzBuzz(-3)).toBe("fizz");
    expect(fizzBuzz(-9)).toBe("fizz");
  });

  it("should handle negative multiples of 5", () => {
    expect(fizzBuzz(-5)).toBe("buzz");
    expect(fizzBuzz(-10)).toBe("buzz");
  });

  it("should handle negative non-multiples", () => {
    expect(fizzBuzz(-1)).toBe("-1");
    expect(fizzBuzz(-7)).toBe("-7");
  });

  it("should handle large numbers", () => {
    expect(fizzBuzz(1000001)).toBe("1000001");
    expect(fizzBuzz(9999999)).toBe("fizz");
  });

  // Boundary cases
  it("should correctly handle 12 (multiple of 3 but not 5)", () => {
    expect(fizzBuzz(12)).toBe("fizz");
  });

  it("should correctly handle 25 (multiple of 5 but not 3)", () => {
    expect(fizzBuzz(25)).toBe("buzz");
  });
});

describe("reverseWords", () => {
  // Happy paths
  it("should reverse the order of words in a simple sentence", () => {
    expect(reverseWords("hello world")).toBe("world hello");
  });

  it("should reverse three words", () => {
    expect(reverseWords("one two three")).toBe("three two one");
  });

  it("should reverse four words", () => {
    expect(reverseWords("the quick brown fox")).toBe("fox brown quick the");
  });

  it("should handle single word", () => {
    expect(reverseWords("hello")).toBe("hello");
  });

  // Edge cases - whitespace
  it("should handle leading whitespace", () => {
    expect(reverseWords("  hello world")).toBe("world hello");
  });

  it("should handle trailing whitespace", () => {
    expect(reverseWords("hello world  ")).toBe("world hello");
  });

  it("should handle multiple spaces between words", () => {
    expect(reverseWords("hello   world")).toBe("world hello");
  });

  it("should handle tabs and mixed whitespace", () => {
    expect(reverseWords("hello\t\nworld")).toBe("world hello");
  });

  it("should handle only whitespace in string", () => {
    expect(reverseWords("   ")).toBe("");
  });

  // Edge cases - empty and special characters
  it("should handle empty string", () => {
    expect(reverseWords("")).toBe("");
  });

  it("should preserve word content with punctuation", () => {
    expect(reverseWords("hello, world!")).toBe("world! hello,");
  });

  it("should handle words with numbers", () => {
    expect(reverseWords("test123 example456")).toBe("example456 test123");
  });

  it("should handle hyphenated words", () => {
    expect(reverseWords("hello-world test")).toBe("test hello-world");
  });

  it("should handle camelCase words", () => {
    expect(reverseWords("camelCase anotherWord")).toBe("anotherWord camelCase");
  });

  // Complex cases
  it("should handle sentence with special characters", () => {
    expect(reverseWords("Hello, World!")).toBe("World! Hello,");
  });

  it("should handle unicode characters", () => {
    expect(reverseWords("café restaurant")).toBe("restaurant café");
  });

  it("should handle many words", () => {
    expect(reverseWords("one two three four five")).toBe(
      "five four three two one"
    );
  });

  it("should maintain word integrity", () => {
    const result = reverseWords("alpha beta gamma");
    const words = result.split(" ");
    expect(words).toEqual(["gamma", "beta", "alpha"]);
  });
});

File: example/go/utils_test.go
// Copyright (c) 2026 Gibran Rodriguez <brangi000@gmail.com>
// SPDX-License-Identifier: MIT

package utils

import (
	"strings"
	"testing"
)

// TestCalculateAverage tests the CalculateAverage function with various inputs
func TestCalculateAverage(t *testing.T) {
	tests := []struct {
		name    string
		numbers []float64
		want    float64
		wantErr bool
	}{
		// Happy paths
		{
			name:    "single number",
			numbers: []float64{5},
			want:    5,
			wantErr: false,
		},
		{
			name:    "two numbers",
			numbers: []float64{2, 4},
			want:    3,
			wantErr: false,
		},
		{
			name:    "multiple numbers",
			numbers: []float64{1, 2, 3, 4, 5},
			want:    3,
			wantErr: false,
		},
		{
			name:    "negative numbers",
			numbers: []float64{-5, -3, -1},
			want:    -3,
			wantErr: false,
		},
		{
			name:    "mixed positive and negative",
			numbers: []float64{-10, 0, 10},
			want:    0,
			wantErr: false,
		},
		{
			name:    "decimal numbers",
			numbers: []float64{1.5, 2.5, 3.5},
			want:    2.5,
			wantErr: false,
		},
		{
			name:    "large numbers",
			numbers: []float64{1e10, 2e10, 3e10},
			want:    2e10,
			wantErr: false,
		},
		{
			name:    "very small numbers",
			numbers: []float64{1e-10, 2e-10, 3e-10},
			want:    2e-10,
			wantErr: false,
		},
		{
			name:    "zeros",
			numbers: []float64{0, 0, 0},
			want:    0,
			wantErr: false,
		},
		// Edge cases
		{
			name:    "empty slice error",
			numbers: []float64{},
			want:    0,
			wantErr: true,
		},
		{
			name:    "nil slice error",
			numbers: nil,
			want:    0,
			wantErr: true,
		},
		{
			name:    "single zero",
			numbers: []float64{0},
			want:    0,
			wantErr: false,
		},
		{
			name:    "many identical numbers",
			numbers: []float64{42, 42, 42, 42},
			want:    42,
			wantErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := CalculateAverage(tt.numbers)

			// Check error condition
			if (err != nil) != tt.wantErr {
				t.Errorf("CalculateAverage() error = %v, wantErr %v", err, tt.wantErr)
				return
			}

			// Check result value (with floating point tolerance)
			if got != tt.want {
				t.Errorf("CalculateAverage() = %v, want %v", got, tt.want)
			}
		})
	}
}

// TestCalculateAveragePrecision tests floating point precision
func TestCalculateAveragePrecision(t *testing.T) {
	// Test that division is accurate
	numbers := []float64{1, 2, 3}
	result, _ := CalculateAverage(numbers)
	expected := 2.0

	if result != expected {
		t.Errorf("CalculateAverage precision test: got %v, want %v", result, expected)
	}
}

// TestFizzBuzz tests the FizzBuzz function with various inputs
func TestFizzBuzz(t *testing.T) {
	tests := []struct {
		name     string
		input    int
		expected string
	}{
		// Happy paths - FizzBuzz numbers
		{
			name:     "15 is fizzbuzz",
			input:    15,
			expected: "fizzbuzz",
		},
		{
			name:     "30 is fizzbuzz",
			input:    30,
			expected: "fizzbuzz",
		},
		{
			name:     "45 is fizzbuzz",
			input:    45,
			expected: "fizzbuzz",
		},
		// Happy paths - Fizz numbers
		{
			name:     "3 is fizz",
			input:    3,
			expected: "fizz",
		},
		{
			name:     "6 is fizz",
			input:    6,
			expected: "fizz",
		},
		{
			name:     "9 is fizz",
			input:    9,
			expected: "fizz",
		},
		{
			name:     "12 is fizz",
			input:    12,
			expected: "fizz",
		},
		// Happy paths - Buzz numbers
		{
			name:     "5 is buzz",
			input:    5,
			expected: "buzz",
		},
		{
			name:     "10 is buzz",
			input:    10,
			expected: "buzz",
		},
		{
			name:     "20 is buzz",
			input:    20,
			expected: "buzz",
		},
		{
			name:     "25 is buzz",
			input:    25,
			expected: "buzz",
		},
		// Happy paths - Regular numbers
		{
			name:     "1 is one",
			input:    1,
			expected: "1",
		},
		{
			name:     "2 is two",
			input:    2,
			expected: "2",
		},
		{
			name:     "4 is four",
			input:    4,
			expected: "4",
		},
		{
			name:     "7 is seven",
			input:    7,
			expected: "7",
		},
		{
			name:     "8 is eight",
			input:    8,
			expected: "8",
		},
		{
			name:     "11 is eleven",
			input:    11,
			expected: "11",
		},
		// Edge cases - Boundary numbers
		{
			name:     "zero is zero",
			input:    0,
			expected: "fizzbuzz", // 0 % 15 == 0
		},
		{
			name:     "negative one",
			input:    -1,
			expected: "-1",
		},
		{
			name:     "negative fizz",
			input:    -3,
			expected: "fizz",
		},
		{
			name:     "negative buzz",
			input:    -5,
			expected: "buzz",
		},
		{
			name:     "negative fizzbuzz",
			input:    -15,
			expected: "fizzbuzz",
		},
		{
			name:     "large number",
			input:    1000000,
			expected: "buzz",
		},
		{
			name:     "large fizzbuzz",
			input:    999990,
			expected: "fizzbuzz",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := FizzBuzz(tt.input)
			if result != tt.expected {
				t.Errorf("FizzBuzz(%d) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

// TestFizzBuzzPattern verifies the priority order (fizzbuzz > fizz > buzz > number)
func TestFizzBuzzPattern(t *testing.T) {
	// Verify that multiples of both 3 and 5 return "fizzbuzz", not "fizz" or "buzz"
	fizzbuzzNumbers := []int{15, 30, 45, 60, 75, 90, 105}
	for _, n := range fizzbuzzNumbers {
		result := FizzBuzz(n)
		if result != "fizzbuzz" {
			t.Errorf("FizzBuzz(%d) = %q, expected fizzbuzz (multiple of both 3 and 5)", n, result)
		}
	}

	// Verify that multiples of only 3 return "fizz"
	fizzNumbers := []int{3, 6, 9, 12, 18, 21, 24}
	for _, n := range fizzNumbers {
		result := FizzBuzz(n)
		if result != "fizz" {
			t.Errorf("FizzBuzz(%d) = %q, expected fizz (multiple of 3 only)", n, result)
		}
	}

	// Verify that multiples of only 5 return "buzz"
	buzzNumbers := []int{5, 10, 20, 25, 35, 40, 50}
	for _, n := range buzzNumbers {
		result := FizzBuzz(n)
		if result != "buzz" {
			t.Errorf("FizzBuzz(%d) = %q, expected buzz (multiple of 5 only)", n, result)
		}
	}
}

// TestReverseWords tests the ReverseWords function with various inputs
func TestReverseWords(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		// Happy paths
		{
			name:     "simple sentence",
			input:    "hello world",
			expected: "world hello",
		},
		{
			name:     "three words",
			input:    "the quick brown",
			expected: "brown quick the",
		},
		{
			name:     "four words",
			input:    "one two three four",
			expected: "four three two one",
		},
		// Edge cases - Single words
		{
			name:     "single word",
			input:    "hello",
			expected: "hello",
		},
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		// Edge cases - Whitespace
		{
			name:     "leading whitespace",
			input:    "  hello world",
			expected: "world hello",
		},
		{
			name:     "trailing whitespace",
			input:    "hello world  ",
			expected: "world hello",
		},
		{
			name:     "multiple spaces between words",
			input:    "hello    world",
			expected: "world hello",
		},
		{
			name:     "tabs and spaces",
			input:    "hello\t\tworld",
			expected: "world hello",
		},
		{
			name:     "mixed whitespace",
			input:    "  hello  \t  world  ",
			expected: "world hello",
		},
		// Edge cases - Special cases
		{
			name:     "only spaces",
			input:    "   ",
			expected: "",
		},
		{
			name:     "only tabs",
			input:    "\t\t\t",
			expected: "",
		},
		{
			name:     "words with punctuation",
			input:    "hello, world!",
			expected: "world! hello,",
		},
		{
			name:     "words with numbers",
			input:    "abc 123 xyz",
			expected: "xyz 123 abc",
		},
		{
			name:     "words with special characters",
			input:    "test@123 sample#456 hello",
			expected: "hello sample#456 test@123",
		},
		// Happy paths - Longer sentences
		{
			name:     "longer sentence",
			input:    "The quick brown fox jumps over lazy dog",
			expected: "dog lazy over jumps fox brown quick The",
		},
		{
			name:     "sentence with repeated words",
			input:    "the the the",
			expected: "the the the",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ReverseWords(tt.input)
			if result != tt.expected {
				t.Errorf("ReverseWords(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

// TestReverseWordsConsistency tests that reversing twice gives original
func TestReverseWordsConsistency(t *testing.T) {
	testCases := []string{
		"hello world",
		"one two three four five",
		"a",
		"test sentence with many words",
	}

	for _, sentence := range testCases {
		t.Run(sentence, func(t *testing.T) {
			reversed := ReverseWords(sentence)
			doubleReversed := ReverseWords(reversed)

			if sentence != doubleReversed {
				t.Errorf("ReverseWords is not consistent: original %q, double reversed %q", sentence, doubleReversed)
			}
		})
	}
}

// TestReverseWordsPreservesContent tests that content is preserved
func TestReverseWordsPreservesContent(t *testing.T) {
	input := "Hello World Test"
	result := ReverseWords(input)
	expected := "Test World Hello"

	if result != expected {
		t.Errorf("ReverseWords(%q) = %q, want %q", input, result, expected)
	}

	// Verify word count is preserved
	inputWords := len(strings.Fields(input))
	resultWords := len(strings.Fields(result))

	if inputWords != resultWords {
		t.Errorf("Word count changed: input %d, result %d", inputWords, resultWords)
	}
}
